The topic of this thesis is reference genomes, and in particular graph-based reference genomes.
In four papers, various aspects of reference genomes are investigated. 
How sequenced reads are mapped to the reference genome, how mapped reads can be used in downstream analysis to deduce biochemical properties of DNA, how results from such analysis can be represented as coordinates and intervals on the reference genome, and how such coordinates and intervals can be used to investigate interplay between different biochemical processes.
Three of the papers are focused on how these processes are performed using a graph based reference genomes.

In the next chapter, brief introductions are given to themes needed to give context to the use of reference genomes in bioinformatics.
Starting out with an introduction to molecular biology in Section~\ref{sec:biology}, Section~\ref{sec:sequencing} explains the process of DNA-sequencing which is how the observed data used in this thesis was obtained.
In Section~\ref{sec:refgenomes} the role of reference genomes is discussed, followed by two sections (\ref{sec:alignment}, \ref{sec:mapping}) on alignment and mapping which introduces some technical aspects of one of the primary functions of reference genomes. Section~\ref{sec:peakcalling} gives some background information on peak calling which is the theme of Paper II. 
Together this background should provide info on what reference genomes are and how they are used, as well as a motivation for using graph based reference genomes.
% 
% 
% erences.
% he provided papers.
% ary understanding of the underlying biology, mathematics and informatics. An brief introduction to the specific subtopics 
% 
% 
%  graph based references.
% 
% lism.
% ain drawbacks of graphical models, namely their complexity.
% rther developed in the subsequent papers.
% ased reference genomes.


%%% Local Variables:
%%% mode: latex
%%% TeX-master: "main"
%%% End:
